File: lintcode/easy/remove_linked_list_elements/py/remove_linked_list_elements.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-linked-list-elements
@Language: Python
@Datetime: 17-02-15 20:05
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param val, an integer
    # @return a ListNode
    def removeElements(self, head, val):
        dummy      = ListNode(0)
        dummy.next = head
        slow       = dummy
        fast       = dummy.next
        
        while fast != None:
            if fast.val != val:
                slow.next = fast
                slow      = slow.next
            fast = fast.next
        
        slow.next = None
            
        return dummy.next

File: lintcode/easy/remove_linked_list_elements/cpp/remove_linked_list_elements.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-linked-list-elements
@Language: C++
@Datetime: 17-02-15 20:08
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head a ListNode
     * @param val an integer
     * @return a ListNode
     */
    ListNode *removeElements(ListNode *head, int val) 
    {
        ListNode 
            dummy{ 0, head },
            *slow{ &dummy },
            *fast{ dummy.next };
            
        while (fast != nullptr) 
        {
            if (fast->val != val) {
                slow->next = fast;
                slow       = slow->next;
            }
            
            fast = fast->next;
        }
        
        slow->next = nullptr;
        
        return dummy.next;
    }
};